### DISCUSSION TITLE
Support for shell-script config lost with Yarn 2?

### DISCUSSION THREAD

**Discussion by ercgrat** [yarnpkg/berry#2589] [Q&A]
In earlier versions of yarn I relied on `yarn config set script-shell <path to shell executable>` to get our repo's scripts to work on windows by using git bash. This configuration option is no longer available, and I can't find an equivalent. How can I configure the default shell for Yarn 2?

**Comment by arcanis** [comment]
You currently can't, because we embed our own shell that ensures the `scripts` commands portability across systems.

What were you using `script-shell` for?

**Reply by ercgrat** [reply]
The repo I'm working in just has a lot of bash scripts in the package.json that don't work using `cmd.exe` on Windows. Setting `script-shell` on npm was not enough to get it to work in Yarn 1, you also had to set it in yarn. This was a long-term solution for me until I just upgraded us to Yarn 2.

That still seems to be the case, or perhaps I need to update some additional configuration to Yarn 2 to get the bash context to work correctly, because it seems to be getting picked up. My understanding of this is pretty limited. I'm just getting errors now related to not understanding windows filepaths, e.g.:
`/bin/bash: C:\Users\ercgr\silo-technologies\silo-web\cypress\support\db\init_be.sh: No such file or directory`

**Reply by ercgrat** [reply]
For context, this is what I referenced months ago to fix yarn 1:
https://stackoverflow.com/a/58188828/1113930

**Reply by merceyz** [reply]
As @arcanis mentioned we ship a shell that should handle the most common stuff on all systems, if you have something that doesn't work please open a feature request (or bug report) and we could look into fixing it.

**Reply by ercgrat** [reply]
Ok I dug into this a little more trying to reproduce for a bug report, and realized that the issue is just the shebang at the top of our script. Before, when I was able to set bash as the default, this didn't cause any issues. Now I have to run yarn starting from a bash context on Windows or I get "No such file or directory" - it was nice when yarn worked from anywhere. This something that has a chance of getting fixed if I write it up?

Specifically, this:
`#!/bin/bash`

**Reply by ArcanisCz** [reply]
Hello, I am facing same problem - I am on Windows and I want yarn scripts to run using git bash for certain project (same as written here for yarn 1.19+ https://stackoverflow.com/questions/23243353/how-to-set-shell-for-npm-run-scripts-in-windows). I have been using similar config for npm for years.

Any chance we can get this config back OR get similar results?

**Reply by clemyan** [reply]
@ArcanisCz As mentioned above, what script are you using that Yarn's portable shell can't handle and *must* be run with bash? Do you have a specific example of such a script?

**Reply by jonsherrard** [reply]
@clemyan pops and pushed commands are used a lot

**Reply by clemyan** [reply]
@jonsherrard
Do you mean `pushd` and `popd`? Are those used a lot **in oneliner `scripts`**?
Instead, you can just save the cwd in a variable: `ORIG_CWD=$(pwd); cd /; pwd; cd $ORIG_CWD; pwd`. Anything more complicated than that should probably be a proper script file instead of a package script.

Not all shells implement `pushd` and `popd` (e.g. `ash` and `dash` don't), and not all platforms have a shell that does. So even without Yarn's portable shell, using `pushd` and `popd` in package scripts is not cross-platform. If you are writing package scripts with shell-specific stuff (e.g. bashisms), then instead of allowing it to silently fail or do unexpected stuff, you should make the script fail-fast by explicitly invoking the desired shell: `bash -c '...'`.

If there is actually a need for `pushd` and `popd`, we can implement those in our portable shell, which would then make them available to all users in a cross-platform manner.
